File: _labs/lab4/R/app-log.R
library(shiny)
library(bslib)
library(bsicons)
library(httr2)
library(logger)
library(jsonlite)

# configure logging ----
logger::log_threshold(level = "INFO")
logger::log_appender(appender = appender_tee(file = "shiny_app.log"))
logger::log_formatter(logger::formatter_glue_or_sprintf)

# session tracking ----
api_url <- "http://127.0.0.1:8080/predict"

ui <- page_sidebar(
  title = "Penguin Mass Predictor",
  theme = bs_theme(bootswatch = "sketchy"),
  
  # logging status indicator ----
  div(
    style = "position: fixed; top: 10px; right: 10px; z-index: 1000;",
    textOutput("log_status", inline = TRUE)
  ),
  
  sidebar = sidebar(
    sliderInput(inputId = "bill_length", 
      label = "Bill Length (mm)",
      min = 30, 
      max = 60, 
      value = 45, 
      step = 1),
    selectInput(inputId = "sex", 
      label = "Sex", 
      choices = c("Male", "Female"), 
      selected = "Male"),
    selectInput(
      inputId = "species",
      label = "Species",
      choices = c("Adelie", "Chinstrap", "Gentoo"),
      selected = "Adelie"),
    actionButton(
      inputId = "predict", 
      label = "Predict", 
      class = "btn-primary")
  ),
  
  # main content -----
  layout_columns(
    card(
      card_header("Penguin Parameters"),
      card_body(
        verbatimTextOutput(outputId = "vals")
      )
    ),
    card(
      card_header("Predicted Mass"),
      card_body(
        value_box(
          showcase_layout = "left center",
          title = "Grams",
          value = textOutput(outputId = "pred"),
          showcase = bs_icon("graph-up"),
          max_height = "200px",
          min_height = "200px",
        )
      )
    ),
    col_widths = c(6, 6),
    row_heights = c(3, 1),
  ),
  
  # system Status ----
  layout_columns(
    card(
      card_header("System Status"),
      card_body(
        h5("API Status:"),
        textOutput("api_health"),
        h5("Recent Logs:"),
        div(
          style = "font-family: 'Courier New', monospace; font-size: 12px; background-color: #f8f9fa; padding: 10px; border-radius: 5px;",
          verbatimTextOutput("recent_logs", placeholder = TRUE)
        ),
        h6(
          "Last updated:", 
            textOutput("log_timestamp", inline = TRUE)
          ),
        style = "max-height: 350px; overflow-y: auto;"
      )
    ),
    # full width for the system status
    col_widths = c(12),
    height = "300px" 
  )
)

server <- function(input, output, session) {
  
  # log app startup within 
  # observer to access session data
  observe({
    logger::log_info("Shiny app started - Session: {session$token} - Host: {session$clientData$url_hostname}")
  }, priority = 1000)  # High priority to run early
  
  # reactive file reader for log monitoring ----
  # polls the log file every 1000ms (1 second) for changes
  log_file_content <- reactiveFileReader(
    intervalMillis = 1000,
    session = session,
    filePath = "shiny_app.log",
    readFunc = function(filePath) {
      if (file.exists(filePath)) {
        # read all lines and get file modification time
        lines <- readLines(filePath, warn = FALSE)
        mod_time <- file.mtime(filePath)
        
        # return both content and timestamp
        list(
          lines = lines,
          last_modified = mod_time,
          total_lines = length(lines)
        )
      } else {
        list(
          lines = character(0),
          last_modified = Sys.time(),
          total_lines = 0
        )
      }
    }
  )
  
  # track user interactions (with throttling)
  observe({
    logger::log_debug("User input changed - Session: {session$token} - bill_length: {input$bill_length} - species: {input$species} - sex: {input$sex}")
  }) |> 
    # throttle to avoid excessive logging
    throttle(2000)  
  
  # reactive values (with validation)
  vals <- reactive({
    bill_length <- input$bill_length
    species <- input$species
    sex <- input$sex
    
    # input validation (with warnings)
    if (bill_length < 30 || bill_length > 60) {
      logger::log_warn("Bill length out of typical range - Session: {session$token} - bill_length: {bill_length}")
    }
    
    if (is.null(species) || is.null(sex)) {
      logger::log_error("Missing required inputs - Session: {session$token} - species_null: {is.null(species)} - sex_null: {is.null(sex)}")
      return(NULL)
    }
    
    data <- data.frame(
      bill_length_mm = bill_length,
      species_Chinstrap = as.numeric(species == "Chinstrap"),
      species_Gentoo = as.numeric(species == "Gentoo"),
      sex_male = as.numeric(sex == "Male")
    )
    
    logger::log_debug("Input data prepared - Session: {session$token} - data: {jsonlite::toJSON(data, auto_unbox = TRUE)}")
    
    return(data)
  })
  
  # API health check
  api_health <- reactive({
    tryCatch({
      logger::log_debug("Checking API health - Session: {session$token}")
      
      response <- httr2::request("http://127.0.0.1:8080/ping") |>
        httr2::req_timeout(5) |>
        httr2::req_perform()
      
      if (httr2::resp_status(response) == 200) {
        logger::log_info("API health check successful - Session: {session$token}")
        return("✅ API Online")
      } else {
        logger::log_warn("API health check returned non-200 status - Session: {session$token} - status: {httr2::resp_status(response)}")
        return("⚠️ API Issues")
      }
    }, error = function(e) {
      logger::log_error("API health check failed - Session: {session$token} - error: {conditionMessage(e)}")
      return("❌ API Offline")
    })
  })
  
  # enhanced prediction with logging
  pred <- reactive({
    request_start <- Sys.time()
    request_data <- vals()
    
    if (is.null(request_data)) {
      logger::log_error("Cannot make prediction with invalid inputs - Session: {session$token}")
      return("❌ Invalid inputs")
    }
    
    logger::log_info("Starting prediction request - Session: {session$token} - request_data: {jsonlite::toJSON(request_data, auto_unbox = TRUE)}")
    
    tryCatch({
      showNotification("Predicting penguin mass...", 
                       type = "default", duration = 3)
      
      response <- httr2::request(api_url) |>
        httr2::req_method("POST") |>
        httr2::req_body_json(request_data, auto_unbox = FALSE) |>
        httr2::req_timeout(30) |>
        httr2::req_perform()
      
      response_time <- as.numeric(difftime(Sys.time(), request_start, units = "secs"))
      response_data <- httr2::resp_body_json(response)
      
      logger::log_info("Prediction successful - Session: {session$token} - response_time_sec: {round(response_time, 3)} - prediction: {response_data$.pred[1]}")
      
      # performance monitoring
      if (response_time > 5) {
        logger::log_warn("Slow API response - Session: {session$token} - response_time_sec: {response_time}")
      }
      
      showNotification("✅ Prediction successful!", 
                       type = "message", duration = 3)
      
      return(response_data$.pred[1])
      
    }, error = function(e) {
      error_msg <- conditionMessage(e)
      response_time <- as.numeric(difftime(Sys.time(), request_start, units = "secs"))
      
      logger::log_error("Prediction request failed - Session: {session$token} - error: {error_msg} - response_time_sec: {round(response_time, 3)}")
      
      # classify error types for better monitoring
      if (grepl("Connection refused|couldn't connect", error_msg, ignore.case = TRUE)) {
        user_msg <- "API not available - is the server running on port 8080?"
        logger::log_error("API connection refused - Session: {session$token}")
      } else if (grepl("timeout|timed out", error_msg, ignore.case = TRUE)) {
        user_msg <- "Request timed out - API may be overloaded"
        logger::log_warn("API timeout occurred - Session: {session$token}")
      } else {
        user_msg <- paste("API Error:", substr(error_msg, 1, 50))
        logger::log_error("Unknown API error - Session: {session$token} - error: {error_msg}")
      }
      
      showNotification(paste("❌", user_msg), type = "error", duration = 5)
      return(paste("❌", user_msg))
    })
  }) |> 
    bindEvent(input$predict, ignoreInit = TRUE)

  # outputs with logging
  output$pred <- renderText({
    prediction <- pred()
    
    if (is.numeric(prediction)) {
      result <- paste(round(prediction, 1), "grams")
      logger::log_info("Displaying prediction to user - Session: {session$token} - display_value: {result}")
      return(result)
    } else {
      logger::log_debug("Displaying error message to user - Session: {session$token} - message: {prediction}")
      return(prediction)
    }
  })
  
  output$vals <- renderPrint({
    data <- vals()
    if (!is.null(data)) {
      logger::log_debug("Displaying input values to user - Session: {session$token}")
      return(data)
    } else {
      return("Invalid inputs")
    }
  })
  
  output$api_health <- renderText({
    api_health()
  })
  
  output$log_status <- renderText({
    paste("Session:", substr(session$token, 1, 8))
  })
  
  # reactive log display that updates when file changes ----
  output$recent_logs <- renderText({
    log_data <- log_file_content()
    
    if (length(log_data$lines) > 0) {
      # get last 5 lines
      recent_lines <- if (log_data$total_lines > 5) {
        tail(log_data$lines, 5)
      } else {
        log_data$lines
      }
      
      # log that we're updating the display (at DEBUG level to avoid spam)
      logger::log_debug("Updating recent logs display - Session: {session$token} - showing {length(recent_lines)} lines")
      
      paste(recent_lines, collapse = "\n")
    } else {
      "No logs available"
    }
  })
  
  # timestamp of last log update ----
  output$log_timestamp <- renderText({
    log_data <- log_file_content()
    format(log_data$last_modified, "%H:%M:%S")
  })
  
  # log session end
  session$onSessionEnded(function() {
    logger::log_info("User session ended - Session: {session$token}")
  })
}

# log app startup at global level
logger::log_info("Shiny application initialized - timestamp: {Sys.time()} - r_version: {R.version.string}")

shinyApp(ui = ui, server = server)

File: _labs/lab4/Python/app-log.py
from shiny import App, render, ui, reactive
import requests
import json
import logging
import time
import os
from datetime import datetime
from pathlib import Path

# configure logging
def setup_logging():
    """Configure logging with file and console output"""
    log_dir = Path("logs")
    log_dir.mkdir(exist_ok=True)
    
    log_file = log_dir / "shiny_app.log"
    
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(levelname)s - %(message)s',
        handlers=[
            logging.FileHandler(log_file, mode='a'),
            logging.StreamHandler()
        ]
    )
    
    return str(log_file)

# initialize logging
log_file_path = setup_logging()
logging.info("Shiny for Python application initialized")

# api configuration
api_url = 'http://127.0.0.1:8080/predict'
ping_url = 'http://127.0.0.1:8080/ping'

app_ui = ui.page_fluid(
    ui.div(
        ui.strong("Session: "),
        ui.output_text(id="session_display"),
        style="position: fixed; top: 10px; right: 10px; z-index: 1000; color: #333; background: #fff; padding: 5px; border-radius: 3px; box-shadow: 0 2px 4px rgba(0,0,0,0.1);"
    ),
    ui.h1("Penguin Mass Predictor"),
    ui.layout_columns(
        ui.card(
            ui.card_header("Inputs"),
            ui.input_slider(id="bill_length", label="Bill Length (mm)", min=30, max=60, value=45, step=0.1),
            ui.input_select(id="sex", label="Sex", choices=["Male", "Female"]),
            ui.input_select(id="species", label="Species", choices=["Adelie", "Chinstrap", "Gentoo"]),
            ui.input_action_button(id="predict", label="Predict")
        ),
        ui.card(
            ui.card_header("Results"),
            ui.h3("Input Values"),
            ui.div(
                ui.output_text("vals_out"),
                style="font-family: 'Monaco', 'Courier New', monospace; font-size: 12px; background-color: #f8f9fa; padding: 10px; border-radius: 5px; max-height: 250px; overflow-y: auto; border: 1px solid #dee2e6; white-space: pre-wrap;"
            ),
            ui.h3("Predicted Mass"),
            ui.div(
                ui.output_text("pred_out"),
                style="font-size: 24px; font-weight: bold; text-align: center; padding: 15px; color: #0066cc;"
            )
        ),
        col_widths=[4, 8]
    ),
    # System Status
    ui.card(
        ui.card_header("System Status"),
        ui.card_body(
            ui.layout_columns(
                ui.div(
                    ui.h4("API Health Check:"),
                    ui.br(),
                    ui.output_text("api_status"),
                    ui.br(),
                    ui.h4("Recent Logs:"),
                    ui.div(
                        ui.output_text("recent_logs_display"),
                        style="font-family: 'Monaco', 'Courier New', monospace; font-size: 12px; background-color: #f8f9fa; padding: 10px; border-radius: 2px; max-height: 250px; overflow-y: auto; border: 2px solid #dee2e6; white-space: pre-wrap;"
                    ),
                    ui.p(
                        "Last updated: ",
                        ui.output_text("log_timestamp"),
                        style="margin-top: 5px; color: #6c757d; font-size: 12px;"
                    )
                ),
                col_widths=[12]
            )
        )
    )
)

def server(input, output, session):
    # generate session ID and log session start
    session_id = f"py_{int(time.time() * 1000) % 100000}"
    logging.info(f"New session started - Session: {session_id}")
    
    # performance tracking reactive values 
    # (keep for error tracking)
    request_times = reactive.value([])
    error_count = reactive.value(0)
    connection_errors = reactive.value(0)
    timeout_errors = reactive.value(0)
    
    # session cleanup
    def on_session_end():
        logging.info(f"Session ended - Session: {session_id}")
    
    session.on_ended(on_session_end)
    
    # input validation with logging
    @reactive.calc
    def vals():
        bill_length = input.bill_length()
        species = input.species()
        sex = input.sex()
        
        # input validation with warnings
        if bill_length < 30 or bill_length > 60:
            logging.warning(f"Bill length out of typical range - Session: {session_id} - bill_length: {bill_length}")
        
        # create data in LIST format 
        # (what API expects)
        d = [{
            "bill_length_mm": float(bill_length),
            "species_Chinstrap": int(species == "Chinstrap"),
            "species_Gentoo": int(species == "Gentoo"),
            "sex_male": int(sex == "Male")
        }]
        
        logging.debug(f"Input data prepared - Session: {session_id} - data: {json.dumps(d)}")
        return d
    
    @reactive.calc
    def api_health_check():
        """Enhanced API health check with logging"""
        try:
            logging.debug(f"Checking API health - Session: {session_id}")
            start_time = time.time()
            
            r = requests.get(ping_url, timeout=5)
            response_time = time.time() - start_time
            
            if r.status_code == 200:
                logging.info(f"API health check successful - Session: {session_id} - response_time: {response_time:.3f}s")
                return f"✅ API is running (ping: {r.json()}) - {response_time:.2f}s"
            else:
                logging.warning(f"API ping failed - Session: {session_id} - status: {r.status_code}")
                return f"⚠️ API ping failed: {r.status_code}"
                
        except requests.exceptions.ConnectionError as e:
            logging.error(f"API connection refused - Session: {session_id} - error: {str(e)}")
            return "❌ Cannot connect to API - is it running on port 8080?"
        except requests.exceptions.Timeout:
            logging.warning(f"API health check timeout - Session: {session_id}")
            return "⚠️ API health check timeout"
        except Exception as e:
            logging.error(f"API health check failed - Session: {session_id} - error: {str(e)}")
            return f"❌ API health check failed: {str(e)}"
    
    @reactive.calc
    @reactive.event(input.predict)
    def pred():
        """Prediction with logging"""
        request_start = time.time()
        data_to_send = vals()
        
        logging.info(f"Starting prediction request - Session: {session_id} - request_data: {json.dumps(data_to_send)}")
        
        try:
            print(f"\n=== PREDICTION REQUEST ===")
            print(f"Sending data to API: {data_to_send}")
            
            r = requests.post(api_url, json=data_to_send, timeout=30)
            response_time = time.time() - request_start
            
            # update performance metrics 
            current_times = request_times()
            current_times.append(response_time)
            if len(current_times) > 10:
                current_times = current_times[-10:]
            request_times.set(current_times)
            
            print(f"HTTP Status Code: {r.status_code}")
            print(f"Raw response text: {r.text}")
            
            if r.status_code == 200:
                result = r.json()
                print(f"✅ Success! Parsed response: {result}")
                
                # handle different possible response formats  
                if '.pred' in result:
                    prediction = result['.pred'][0]
                elif 'predict' in result:
                    prediction = result['predict'][0]
                else:
                    logging.warning(f"Unexpected response format - Session: {session_id} - response: {result}")
                    return f"Unexpected response format: {result}"
                
                logging.info(f"Prediction successful - Session: {session_id} - response_time: {response_time:.3f}s - prediction: {prediction}")
                
                # performance warning
                if response_time > 5:
                    logging.warning(f"Slow API response - Session: {session_id} - response_time: {response_time:.3f}s")
                
                return prediction
            else:
                error_msg = f"API Error {r.status_code}: {r.text}"
                logging.error(f"Prediction request failed - Session: {session_id} - status: {r.status_code} - response: {r.text[:200]}")
                error_count.set(error_count() + 1)
                return error_msg
                
        except requests.exceptions.ConnectionError as e:
            error_msg = f"Connection Error: {str(e)}"
            logging.error(f"API connection refused during prediction - Session: {session_id} - error: {str(e)}")
            connection_errors.set(connection_errors() + 1)
            error_count.set(error_count() + 1)
            print(f"❌ Connection Error: {e}")
            return error_msg
        except requests.exceptions.Timeout:
            error_msg = "Request timed out - API may be overloaded"
            logging.warning(f"API timeout during prediction - Session: {session_id}")
            timeout_errors.set(timeout_errors() + 1)
            error_count.set(error_count() + 1)
            return error_msg
        except Exception as e:
            error_msg = f"Error: {str(e)}"
            logging.error(f"Unknown prediction error - Session: {session_id} - error: {str(e)}")
            error_count.set(error_count() + 1)
            print(f"❌ Error: {e}")
            return error_msg

    # log file monitoring
    @reactive.file_reader(log_file_path)
    def log_file_content():
        """Monitor log file for changes"""
        try:
            with open(log_file_path, 'r') as f:
                lines = f.readlines()
                return {
                    'lines': lines,
                    'last_modified': datetime.now(),
                    'total_lines': len(lines)
                }
        except Exception as e:
            logging.error(f"Error reading log file - Session: {session_id} - error: {str(e)}")
            return {
                'lines': [],
                'last_modified': datetime.now(),
                'total_lines': 0
            }

    # output renderers -------
    @render.text
    def session_display():
        return session_id[:8]

    @render.text
    def vals_out():
        data = vals()
        logging.debug(f"Displaying input values to user - Session: {session_id}")
        return f"{data}"
    
    @render.text
    def api_status():
        return api_health_check()

    @render.text
    def pred_out():
        result = pred()
        if isinstance(result, (int, float)):
            display_value = f"{round(result, 1)} grams"
            logging.info(f"Displaying prediction to user - Session: {session_id} - display_value: {display_value}")
            return display_value
        else:
            return str(result)
    
    @render.text
    def recent_logs_display():
        log_data = log_file_content()
        lines = log_data['lines']
        
        if lines:
            # get last 10 lines for better monitoring
            recent_lines = lines[-10:] if len(lines) > 10 else lines
            clean_lines = [line.rstrip() for line in recent_lines]
            
            logging.debug(f"Updating recent logs display - Session: {session_id} - showing {len(recent_lines)} lines")
            return '\n'.join(clean_lines)
        else:
            return "No logs available yet..."
    
    @render.text
    def log_timestamp():
        log_data = log_file_content()
        return log_data['last_modified'].strftime("%Y-%m-%d %H:%M:%S")

app = App(app_ui, server)

# app creation
logging.info("Shiny for Python application created successfully")